### DISCUSSION TITLE
I implement a specific IActivityExecutionMiddleware and configure it in Program.cs by Elsa3.0,but I have a few questions.

### DISCUSSION THREAD

**Discussion by coolshun** [elsa-workflows/elsa-core#5193] [Q&A]
I implement a specific IActivityExecutionMiddleware and configure it in Program.cs by Elsa3.0,but I have a few questions below.
1,Why is there no "Completed" status for "Workflow Id:e9ad832c9c1a83a6  Activity.Id:Workflow1" in "TestWorkflowMiddleware"?I can't get the Completed status in business system.
2,Why is there no "Completed" status for " Activity Id:6691cdf6bb50e033  NodeId:Workflow1 Activity.Id:Workflow1" in "TestActivityMiddleware"?
3,Why is there no "Completed" status for " Activity  Id:3fcb1fbbe8d5eafd  NodeId:Workflow1:163b569f74c5f273  Activity.Id:163b569f74c5f273" in "TestActivityMiddleware"?this is Elsa.Flowchart  root node.

The code below:
TestActivityMiddleware.cs:
`
using Elsa.Workflows.Contracts;
using Elsa.Workflows.Pipelines.ActivityExecution;
using Elsa.Workflows;

namespace ElsaServer.Middleware
{
    /// <summary>
    /// Provides extension methods to <see cref="IActivityExecutionPipelineBuilder"/>.
    /// </summary>
    public static class TestActivityMiddlewareExtensions
    {
        /// <summary>
        /// Adds the <see cref="TestActivityMiddleware"/> component to the pipeline.
        /// </summary>
        public static IActivityExecutionPipelineBuilder UseTestActivityMiddleware(this IActivityExecutionPipelineBuilder pipelineBuilder) => pipelineBuilder.UseMiddleware<TestActivityMiddleware>();
    }

    /// <summary>
    /// A default activity execution middleware component that evaluates the current activity's properties, executes the activity and adds any produced bookmarks to the workflow execution context.
    /// </summary>
    public class TestActivityMiddleware : IActivityExecutionMiddleware
    {
        private readonly ActivityMiddlewareDelegate _next;

        /// <summary>
        /// Constructor
        /// </summary>
        public TestActivityMiddleware(ActivityMiddlewareDelegate next)
        {
            _next = next;
        }

        /// <inheritdoc />
        public async ValueTask InvokeAsync(ActivityExecutionContext context)
        {
            Console.WriteLine($"Before Activity Execution: Id:{context.Id}  NodeId:{context.NodeId} Status:{context.Status} Activity.Id:{context.Activity.Id}   Activity.Name:{context.Activity.Name}");
            // Invoke next middleware.
            await _next(context);
            Console.WriteLine($"After Activity Execution: Id:{context.Id}  NodeId:{context.NodeId} Status:{context.Status} Activity.Id:{context.Activity.Id}   Activity.Name:{context.Activity.Name}");
        }

    }
}
`

TestWorkflowMiddleware.cs:
`
using Elsa.Workflows.Contracts;
using Elsa.Workflows.Pipelines.WorkflowExecution;
using Elsa.Workflows;

namespace ElsaServer.Middleware
{

    /// <summary>
    /// Installs middleware that executes scheduled work items.
    /// </summary>
    public static class UseTestWorkflowMiddlewareExtensions
    {
        /// <summary>
        /// Installs middleware that executes scheduled work items. 
        /// </summary>
        public static IWorkflowExecutionPipelineBuilder UseTestWorkflowMiddleware(this IWorkflowExecutionPipelineBuilder pipelineBuilder) => pipelineBuilder.UseMiddleware<TestWorkflowMiddleware>();
    }

    /// <summary>
    /// A workflow execution middleware component that executes scheduled work items.
    /// </summary>
    public class TestWorkflowMiddleware : WorkflowExecutionMiddleware
    {
        private readonly IActivityInvoker _activityInvoker;

        /// <inheritdoc />
        public TestWorkflowMiddleware(WorkflowMiddlewareDelegate next, IActivityInvoker activityInvoker) : base(next)
        {
            _activityInvoker = activityInvoker;
        }

        /// <inheritdoc />
        public override async ValueTask InvokeAsync(WorkflowExecutionContext context)
        {
            Console.WriteLine($"Before Workflow Id:{context.Id}   Status:{context.Status} Activity.Id:{context.Activity.Id}   Activity.Name:{context.Activity.Name}");
            await Next(context);
            Console.WriteLine($"After Workflow Id:{context.Id}   Status:{context.Status} Activity.Id:{context.Activity.Id}   Activity.Name:{context.Activity.Name}");
        }

    }
}

`
Program.cs:
`

using Elsa.EntityFrameworkCore.Modules.Management;
using Elsa.EntityFrameworkCore.Modules.Runtime;
using Elsa.Extensions;

using Elsa.Webhooks.Extensions;
using Elsa.Workflows.Contracts;
using Elsa.Workflows.Middleware.Activities;
using Elsa.Workflows.Middleware.Workflows;
using ElsaServer.Middleware;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddElsa(elsa =>
{
    // Configure Management layer to use EF Core.
    elsa.UseWorkflowManagement(management => management.UseEntityFrameworkCore());

    // Configure Runtime layer to use EF Core.
    elsa.UseWorkflowRuntime(runtime => runtime.UseEntityFrameworkCore());

    // Default Identity features for authentication/authorization.
    elsa.UseIdentity(identity =>
    {
        identity.TokenOptions = options => options.SigningKey = "sufficiently-large-secret-signing-key"; // This key needs to be at least 256 bits long.
        identity.UseAdminUserProvider();
    });

    // Configure ASP.NET authentication/authorization.
    elsa.UseDefaultAuthentication(auth => auth.UseAdminApiKey());

    // Expose Elsa API endpoints.
    elsa.UseWorkflowsApi();

    // Setup a SignalR hub for real-time updates from the server.
    elsa.UseRealTimeWorkflows();

    // Enable JavaScript workflow expressions
    elsa.UseJavaScript(options => options.AllowClrAccess = true);

    // Enable C# workflow expressions
    elsa.UseCSharp();

    // Enable HTTP activities.
    elsa.UseHttp();

    // Use timer activities.
    elsa.UseScheduling();

    // Use email activities.
    elsa.UseEmail(email =>
    {
        email.ConfigureOptions = options =>
        {
            options.Host = "localhost";
            options.Port = 2525;
        };
    });

    // Register custom webhook definitions from the application, if any.
    elsa.UseWebhooks(webhooks => webhooks.WebhookOptions = options => builder.Configuration.GetSection("Webhooks").Bind(options));

    // Register custom activities from the application, if any.
    elsa.AddActivitiesFrom<Program>();

    // Register custom workflows from the application, if any.
    elsa.AddWorkflowsFrom<Program>();

    ////elsa.AddActivity<Greeter>();
    elsa.UseWorkflows(workflows =>
     {
         // Configure workflow execution pipeline to handle workflow contexts.
         workflows.WithWorkflowExecutionPipeline(pipeline => pipeline
             .Reset()
             .UsePersistentVariables()
             .UseBookmarkPersistence()
             .UseWorkflowExecutionLogPersistence()
             .UseWorkflowExecutionLogPersistence()
             .UseActivityExecutionLogPersistence()
             .UseWorkflowContexts()
             .UseExceptionHandling()
             .UseDefaultActivityScheduler()
             .UseTestWorkflowMiddleware()
         );

         // Configure activity execution pipeline to handle workflow contexts.
         workflows.WithActivityExecutionPipeline(pipeline => pipeline
             .Reset()
             .UseExceptionHandling()
             .UseExecutionLogging()
             .UseWorkflowContexts()
             .UseBackgroundActivityInvoker()
             .UseTestActivityMiddleware()
         );
     });
});
//builder.Services.AddScoped<IActivityExecutionMiddleware, TestActivityMiddleware>();
// Configure CORS to allow designer app hosted on a different origin to invoke the APIs.
builder.Services.AddCors(cors => cors
    .AddDefaultPolicy(policy => policy
        .AllowAnyOrigin() // For demo purposes only. Use a specific origin instead.
        .AllowAnyHeader()
        .AllowAnyMethod()
        .WithExposedHeaders("x-elsa-workflow-instance-id"))); // Required for Elsa Studio in order to support running workflows from the designer. Alternatively, you can use the `*` wildcard to expose all headers.

// Add Health Checks.
builder.Services.AddHealthChecks();

// Build the web application.
var app = builder.Build();

// Configure web application's middleware pipeline.
app.UseCors();
app.UseRouting(); // Required for SignalR.
app.UseAuthentication();
app.UseAuthorization();
app.UseWorkflowsApi(); // Use Elsa API endpoints.
app.UseWorkflows(); // Use Elsa middleware to handle HTTP requests mapped to HTTP Endpoint activities.
app.UseWorkflowsSignalRHubs(); // Optional SignalR integration. Elsa Studio uses SignalR to receive real-time updates from the server. 

app.Run();
`

Workflow Definition Json:
`
{
    "id": "8eec3cafafd89cbc",
    "definitionId": "61f466b0e15da281",
    "name": "test single RunTask",
    "createdAt": "2024-03-22T04:56:52.4906207+00:00",
    "version": 1,
    "toolVersion": "3.0.0.0",
    "variables": [],
    "inputs": [],
    "outputs": [],
    "outcomes": [],
    "customProperties": {},
    "isReadonly": false,
    "isLatest": true,
    "isPublished": false,
    "options": {
        "autoUpdateConsumingWorkflows": false
    },
    "root": {
        "type": "Elsa.Flowchart",
        "version": 1,
        "id": "163b569f74c5f273",
        "nodeId": "Workflow1:163b569f74c5f273",
        "metadata": {},
        "customProperties": {
            "source": "FlowchartJsonConverter.cs:45",
            "notFoundConnections": [],
            "canStartWorkflow": false,
            "runAsynchronously": false
        },
        "activities": [
            {
                "taskName": {
                    "typeName": "String",
                    "expression": {
                        "type": "Literal",
                        "value": "TestTask"
                    }
                },
                "payload": {
                    "typeName": "ObjectDictionary",
                    "expression": {
                        "type": "JavaScript",
                        "value": "return {\r\n    employee:{Name:\"Alice Smith\",Email:\"alice.smith@acme.com\"},\r\n    description:\"workflow complete\"\r\n}"
                    }
                },
                "result": null,
                "id": "c381619bcf0842c",
                "nodeId": "Workflow1:163b569f74c5f273:c381619bcf0842c",
                "name": "RunTask1",
                "type": "Elsa.RunTask",
                "version": 1,
                "customProperties": {
                    "canStartWorkflow": false,
                    "runAsynchronously": false
                },
                "metadata": {
                    "designer": {
                        "position": {
                            "x": -129.4761962890625,
                            "y": 320.4761962890625
                        },
                        "size": {
                            "width": 132.17857360839844,
                            "height": 49.52381134033203
                        }
                    }
                }
            },
            {
                "id": "cd2c745f5c784913",
                "nodeId": "Workflow1:163b569f74c5f273:cd2c745f5c784913",
                "name": "End1",
                "type": "Elsa.End",
                "version": 1,
                "customProperties": {
                    "canStartWorkflow": false,
                    "runAsynchronously": false
                },
                "metadata": {
                    "designer": {
                        "position": {
                            "x": 102.13690185546875,
                            "y": 320.4761962890625
                        },
                        "size": {
                            "width": 93.5,
                            "height": 49.52381134033203
                        }
                    }
                }
            }
        ],
        "connections": [
            {
                "source": {
                    "activity": "c381619bcf0842c",
                    "port": "Done"
                },
                "target": {
                    "activity": "cd2c745f5c784913",
                    "port": "In"
                }
            }
        ]
    }
}
`

The Log Output:
`
 Before Activity Execution: Id:6691cdf6bb50e033  NodeId:Workflow1 Status:Running Activity.Id:Workflow1   Activity.Name:
 After Activity Execution: Id:6691cdf6bb50e033  NodeId:Workflow1 Status:Running Activity.Id:Workflow1   Activity.Name:
 Before Activity Execution: Id:3fcb1fbbe8d5eafd  NodeId:Workflow1:163b569f74c5f273 Status:Running Activity.Id:163b569f74c5f273   Activity.Name:
 After Activity Execution: Id:3fcb1fbbe8d5eafd  NodeId:Workflow1:163b569f74c5f273 Status:Running Activity.Id:163b569f74c5f273   Activity.Name:
 Before Activity Execution: Id:35759ef220a9c71f  NodeId:Workflow1:163b569f74c5f273:c381619bcf0842c Status:Running Activity.Id:c381619bcf0842c   Activity.Name:RunTask1
 After Activity Execution: Id:35759ef220a9c71f  NodeId:Workflow1:163b569f74c5f273:c381619bcf0842c Status:Running Activity.Id:c381619bcf0842c   Activity.Name:RunTask1
 Before Workflow Id:e9ad832c9c1a83a6   Status:Running Activity.Id:Workflow1   Activity.Name:
 After Workflow Id:e9ad832c9c1a83a6   Status:Running Activity.Id:Workflow1   Activity.Name:

 Before Activity Execution: Id:35759ef220a9c71f  NodeId:Workflow1:163b569f74c5f273:c381619bcf0842c Status:Completed Activity.Id:c381619bcf0842c   Activity.Name:RunTask1
 After Activity Execution: Id:35759ef220a9c71f  NodeId:Workflow1:163b569f74c5f273:c381619bcf0842c Status:Completed Activity.Id:c381619bcf0842c   Activity.Name:RunTask1
 Before Activity Execution: Id:558246fe87533df3  NodeId:Workflow1:163b569f74c5f273:cd2c745f5c784913 Status:Completed Activity.Id:cd2c745f5c784913   Activity.Name:End1
 After Activity Execution: Id:558246fe87533df3  NodeId:Workflow1:163b569f74c5f273:cd2c745f5c784913 Status:Completed Activity.Id:cd2c745f5c784913   Activity.Name:End1
 Before Workflow Id:e9ad832c9c1a83a6   Status:Running Activity.Id:Workflow1   Activity.Name:
 After Workflow Id:e9ad832c9c1a83a6   Status:Running Activity.Id:Workflow1   Activity.Name:
`

**Comment by glime-ai** [comment]
Based on the information provided, it seems you are trying to capture the "Completed" status of activities and workflows within your custom middleware in an Elsa 3.0 workflow application. However, you are not seeing the "Completed" status in your logs for certain activities and the workflow itself.

1. The absence of the "Completed" status for "Workflow Id:e9ad832c9c1a83a6 Activity.Id:Workflow1" in "TestWorkflowMiddleware" could be due to the middleware not being invoked after the workflow has completed. The middleware pipeline is designed to handle the execution of activities and workflows, but it may not be capturing the final state after all activities have been executed. To capture the completed status, you might need to subscribe to workflow-related events or notifications that Elsa publishes when workflows change their status.

2. Similarly, for "Activity Id:6691cdf6bb50e033 NodeId:Workflow1 Activity.Id:Workflow1" and "Activity Id:3fcb1fbbe8d5eafd NodeId:Workflow1:163b569f74c5f273 Activity.Id:163b569f74c5f273," the middleware may not be capturing the final state of the activities. You may need to implement additional logic to handle the completion event of activities.

3. For the Elsa.Flowchart root node, the same reasoning applies. The middleware may not be capturing the final state, and you may need to handle the completion event specifically.

To address these issues, you can explore the following options:

- Subscribe to workflow and activity events: Elsa 3.0 provides a mechanism to subscribe to various workflow and activity events. You can use this feature to listen for the completion of workflows and activities and then perform your custom logic.
- Review the middleware pipeline: Ensure that your middleware is correctly configured in the pipeline and that it is capable of capturing the final states of workflows and activities.
- Debug the application: Use debugging tools to step through the execution of your middleware and the workflow to understand where the "Completed" status might be getting lost.

If you need further assistance, consider reaching out to the Elsa community or reviewing the Elsa documentation for more details on how to work with middleware and capture workflow and activity events.

**Comment by glime-ai** [comment]
Based on the information provided, it seems that you are not seeing the "Completed" status for certain activities and workflows when using custom middleware in Elsa 3.0. This could be due to several reasons, including but not limited to:

1. The middleware may not be correctly configured to handle the completion status.
2. The workflow or activities might not be reaching a completion state due to some error or misconfiguration.
3. The logging statements might not be capturing the "Completed" status due to their placement in the code.

To troubleshoot this issue, you can consider the following steps:

- Ensure that the middleware is correctly configured and that it is part of the pipeline.
- Verify that the workflow and activities are correctly defined and that they are capable of reaching a completion state.
- Check if there are any errors or exceptions being thrown that might prevent the workflow or activities from completing.
- Place logging statements at strategic points in the code to capture the status changes, especially after the activity or workflow has been executed.

If you continue to face issues, you might want to reach out to the Elsa community or check the official documentation for more guidance on how to implement and configure custom middleware correctly.

Since you are using Elsa 3.0, you can also look into the possibility of subscribing to events or notifications that Elsa might publish when activities change their status. This could provide an alternative way to monitor the status of your workflows and activities.

For more detailed assistance, you may need to provide additional context or code snippets, or consider posting a question on platforms like Stack Overflow or the Elsa discussion forums where the community can provide more targeted support.
